docs/building-modules/upgrade.md

```diff
@@ -0,0 +1,55 @@
+
+```golang
+func (am AppModule) RegisterServices(cfg module.Configurator) {
+    // --snip--
+    cfg.RegisterMigration(types.ModuleName, 1, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 1 to 2.
+    })
+     cfg.RegisterMigration(types.ModuleName, 2, func(ctx sdk.Context) error {
+        // Perform in-place store migrations from ConsensusVersion 2 to 3.
+    })
+}
+```
+
+Since these migrations are functions that need access to a Keeper's store, we use a wrapper around the keepers called `Migrator`. An example of this can be found here:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/6ac8898fec9bd7ea2c1e5c79e0ed0c3f827beb55/x/bank/keeper/migrations.go#L8-L21
+
+In addition to the `Migrator` wrapper, we also define our migration scripts. More on that below.
+
+## Writing Migration Scripts
+
+In order to define the functionality that takes place during an upgrade, we will write a migration script. Since migration scripts will manipulate legacy code, we place these functions in a `legacy/` directory. For example, if we wanted to write migration scripts for a module named `bank`, we would place the functions in `x/bank/legacy/`. We recommend the following naming convention for these functions:
+
+```golang
+// Migrating bank module from version 1 to 2
+func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+	return v043bank.MigrateStore(ctx, m.keeper.storeKey) // v043bank is package `x/bank/legacy/v043`.
+}
+```
+
+If you would like to see example code of changes implemented in a migration, you can check out the code [here](https://github.com/cosmos/cosmos-sdk/blob/36f68eb9e041e20a5bb47e216ac5eb8b91f95471/x/bank/legacy/v043/store.go#L41-L62). For context, this introduced migrations of the bank store that updated addresses to be prefixed by their length in bytes as perscribed in (ADR-028)[../architecture/adr-028-public-key-addresses.md].
+
+
+## Adding New Modules In Upgrades
+
+When introducing a new module to your application during an upgrade, your `AppModule` must begin at consensus version 0. When an upgrade is being executed, all modules with consensus versions set to 0 will be recognized as new modules. This will result in the new module's `InitGenesis` function executing. This allows new modules to setup some inital state. The result of this will set the `AppModule`s consensus version to `1`.
+
+If your module does not require any initial state, you do not need to implement any `InitGenesis`. The `InitGenesis` will result in a no-op, but still upgrade your consensus version to `1` for future upgrades. 
```

Comment:
I would also add a paragraph like this:

---

If you wish to overwrite the default behavior of running InitGenesis during an upgrade for new modules, make sure to pass the latest `ConsensusVersion` of the new module into the return `module.VersionMap`. This will then skip running InitGenesis for the module:

```go
// Let's say foo is a new module
import foo "github.com/my/module/foo"

app.UpgradeKeeper.SetUpgradeHandler("my-plan", func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap)  (module.VersionMap, error) {
    // We make sure to set foo's version to the latest ConsensusVersion in the VersionMap.
    // This will skip running InitGenesis on Foo
    vm["foo"] = foo.AppModule{}.ConsensusVersion()

    return app.mm.RunMigrations(ctx, vm)
})
```

Using the same method, you can also run InitGenesis on your new module with a custom genesis state:

```go
app.UpgradeKeeper.SetUpgradeHandler("my-plan", func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap)  (module.VersionMap, error) {
    vm["foo"] = foo.AppModule{}.ConsensusVersion()

    // Run custom InitGenesis for foo
    app.mm["foo"].InitGenesis(ctx, app.appCodec, myCustomGenesisState)

    return app.mm.RunMigrations(ctx, vm)
})
```